src/node_dns.cc

```diff
@@ -0,0 +1,471 @@
+#include "node_dns.h"
+#include "base_object-inl.h"
+#include "env-inl.h"
+#include "node.h"
+#include "node_errors.h"
+#include "node_mem-inl.h"
+#include "util-inl.h"
+#include "uv.h"
+
+namespace node {
+namespace dns {
+
+using v8::Array;
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Global;
+using v8::HandleScope;
+using v8::Integer;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Name;
+using v8::NewStringType;
+using v8::Null;
+using v8::Object;
+using v8::Promise;
+using v8::String;
+using v8::Value;
+
+#define GETDNS_BASE(ret, name, isolate, ...)                                   \
+  do {                                                                         \
+    getdns_return_t r = getdns_##name(__VA_ARGS__);                            \
+    if (UNLIKELY(r != GETDNS_RETURN_GOOD)) {                                   \
+      THROW_ERR_DNS_ERROR(isolate, getdns_get_errorstr_by_id(r));              \
+      return ret;                                                              \
+    }                                                                          \
+  } while (false);
+
+#define GETDNS(name, isolate, ...)                                             \
+  GETDNS_BASE(/* empty */, name, isolate, __VA_ARGS__)
+#define GETDNSV(name, isolate, ...) GETDNS_BASE({}, name, isolate, __VA_ARGS__)
+
+DNSWrap::DNSWrap(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {
+  Isolate* isolate = env->isolate();
+  allocator_ = MakeAllocator();
+  GETDNS(context_create_with_extended_memory_functions,
+         isolate,
+         &context_,
+         true,
+         reinterpret_cast<void*>(&allocator_),
+         [](void* data, size_t size) -> void* {
+           Allocator* a = reinterpret_cast<Allocator*>(data);
+           return a->malloc(size, a->data);
+         },
+         [](void* data, void* ptr, size_t size) -> void* {
+           Allocator* a = reinterpret_cast<Allocator*>(data);
+           return a->realloc(ptr, size, a->data);
+         },
+         [](void* data, void* ptr) {
+           Allocator* a = reinterpret_cast<Allocator*>(data);
+           a->free(ptr, a->data);
+         });
+  GETDNS(extension_set_libuv_loop, isolate, context_, env->event_loop());
+  GETDNS(context_set_tls_authentication,
+         isolate,
+         context_,
+         GETDNS_AUTHENTICATION_REQUIRED);
+}
+
+DNSWrap::~DNSWrap() {
+  getdns_context_destroy(context_);
```

Comment:
adding a `using GetDnsContextPointer = DeleteFnPtr<getdns_context, getdns_context_destroy>;` would be nice.